core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -642,9 +679,99 @@ public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<
             lock.writeLock().unlock();
         }
 
+        if (canMoveStartOffset) {
+            maybeUpdateCachedStateAndOffsets();
+        }
+
         return CompletableFuture.completedFuture(Optional.ofNullable(throwable));
     }
 
+    // Visible for testing
+    public long startOffset() {
+        return this.startOffset;
+    }
+
+    // Visible for testing
+    public long endOffset() {
+        return this.endOffset;
+    }
+
+    public boolean canAcquireMore() {
+        lock.readLock().lock();
+        long numRecords = this.endOffset - this.startOffset + 1;
+        lock.readLock().unlock();
+        return numRecords < recordLockPartitionLimit;
+    }
+
+    public void maybeUpdateCachedStateAndOffsets() {
+        lock.writeLock().lock();
```

Comment:
You should use this lock in a try-finally block so we know the lock is released no matter how the method exits (exceptions included).